# Exploring eBay Used Car Sales Data

### Intro 

In this project, we'll be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The aim of this project is to clean the data and analyze the included used car listings and try to do the following analysis:

- Find the most common brand/model combinations
- Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
- How much cheaper are cars with damage than their non-damaged counterparts?

The set contains the following columns.

- dateCrawled: When the ad was first crawled. All other field values for the - corresponding row were scraped on this date.
- name: Name of the car listing.
- seller: Whether the seller is a private owner or a dealer.
- offerType: The type of listing.
- price: The price on the ad to sell the car.
- abtest: Whether the listing is included in an A/B test.
- vehicleType: The vehicle type.
- yearOfRegistration: The year in which the car was first registered.
- gearbox: The transmission type.
- powerPS: The power of the car in PS.
- model: The car model name.
- odometer: The odometer reading on the car, in kilometers.
- monthOfRegistration: The month in which the car was first registered.
- fuelType: What type of fuel the car uses.
- brand: The brand of the car.
- notRepairedDamage: Whether or not the car has damage which is not yet repaired.
- dateCreated: The date on which the eBay listing was created.
- nrOfPictures: The number of pictures in the listing.
- postalCode: The postal code for the location of the vehicle.
- lastSeenOnline: When the crawler last saw this listing online.

The original dataset can be found [here](https://data.world/data-society/used-cars-data).

### Read Data into Dataframe

In [1]:
import pandas as pd #importing pandas library
import numpy as np  #importing numpy library
autos = pd.read_csv('autos.csv',encoding = 'Latin-1')

In [2]:
autos

dateCrawled                                              name  \
0      2016-03-24 11:52:17                                        Golf_3_1.6   
1      2016-03-24 10:58:45                              A5_Sportback_2.7_Tdi   
2      2016-03-14 12:52:21                    Jeep_Grand_Cherokee_"Overland"   
3      2016-03-17 16:54:04                               GOLF_4_1_4__3TÜRER   
4      2016-03-31 17:25:20                    Skoda_Fabia_1.4_TDI_PD_Classic   
...                    ...                                               ...   
42744  2016-03-16 16:06:20                             Alfa_Romeo_4C_1.8_TBi   
42745  2016-03-25 08:56:51                          BMW_X6_xDrive30d__1_HAND   
42746  2016-03-31 07:36:22                                          BMW_118i   
42747  2016-03-31 18:57:12  Mercedes_Benz_E_55_AMG_T_~19_Zoll~V_max~Tuev_Neu   
42748  2016-03-28 02:00:51     Mitsubishi_Lancer_Kombi_1.6_Comfort_AUTOMATIK   

       seller offerType  price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot    480     test         NaN                1993   
1      privat   Angebot  18300     test       coupe                2011   
2      privat   Angebot   9800     test         suv                2004   
3      privat   Angebot   1500     test  kleinwagen                2001   
4      privat   Angebot   3600     test  kleinwagen                2008   
...       ...       ...    ...      ...         ...                 ...   
42744  privat   Angebot  56900     test       coupe                2016   
42745  privat   Angebot  31200  control   limousine                2009   
42746  privat   Angebot   9999     test   limousine                2008   
42747  privat   Angebot  15800  control       kombi                2003   
42748  privat   Angebot   1750     test       kombi                2004   

         gearbox  powerPS     model  kilometer  monthOfRegistration fuelType  \
0        manuell        0      golf     150000                    0   benzin   
1        manuell      190       NaN     125000                    5   diesel   
2      automatik      163     grand     125000                    8   diesel   
3        manuell       75      golf     150000                    6   benzin   
4        manuell       69     fabia      90000                    7   diesel   
...          ...      ...       ...        ...                  ...      ...   
42744  automatik      241    andere       5000                    3   benzin   
42745  automatik      235   x_reihe      80000                    7   diesel   
42746    manuell      143       1er      70000                    9   benzin   
42747  automatik      500  e_klasse     150000                    9   benzin   
42748  automatik       98    lancer     150000                    8       be   

               brand notRepairedDamage          dateCreated  nrOfPictures  \
0         volkswagen               NaN  2016-03-24 00:00:00           0.0   
1               audi                ja  2016-03-24 00:00:00           0.0   
2               jeep               NaN  2016-03-14 00:00:00           0.0   
3         volkswagen              nein  2016-03-17 00:00:00           0.0   
4              skoda              nein  2016-03-31 00:00:00           0.0   
...              ...               ...                  ...           ...   
42744     alfa_romeo              nein  2016-03-16 00:00:00           0.0   
42745            bmw              nein  2016-03-25 00:00:00           0.0   
42746            bmw              nein  2016-03-31 00:00:00           0.0   
42747  mercedes_benz              nein  2016-03-31 00:00:00           0.0   
42748            NaN               NaN                  NaN           NaN   

       postalCode             lastSeen  
0         70435.0  2016-04-07 03:16:57  
1         66954.0  2016-04-07 01:46:50  
2         90480.0  2016-04-05 12:47:46  
3         91074.0  2016-03-17 17:40:17  
4         60437.0  2016-04-06 10:17:21  
...           ...    

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42749 entries, 0 to 42748
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   dateCrawled          42749 non-null  object 
 1   name                 42749 non-null  object 
 2   seller               42749 non-null  object 
 3   offerType            42749 non-null  object 
 4   price                42749 non-null  int64  
 5   abtest               42749 non-null  object 
 6   vehicleType          38300 non-null  object 
 7   yearOfRegistration   42749 non-null  int64  
 8   gearbox              40420 non-null  object 
 9   powerPS              42749 non-null  int64  
 10  model                40381 non-null  object 
 11  kilometer            42749 non-null  int64  
 12  monthOfRegistration  42749 non-null  int64  
 13  fuelType             38786 non-null  object 
 14  brand                42748 non-null  object 
 15  notRepairedDamage    34413 non-null 

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04             GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00           0.0     70435.0  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0.0     66954.0  2016-04-07 01:46:50  
2  2016-03-14 00:00:00           0.0     90480.0  2016-04-05 12:47:46  
3  2016-03-17 00:00:00           0.0     91074.0  2016-03-17 17:40:17  
4  2016-03-31 00:00:00           0.0     60437.0  2016-04-06 10:17:21

From the cell above, we see that the dataset contains 50,000 rows and 20 columns. Most of the columns don't have missing values except for the **'vehicletype'**, **'gearbox'**, **'model'**, **'fueltype'** and **'notRepairedDamage'** columns. 12 columns consists as $\color{green}{\text{strings}}$, 5 $\color{green}{\text{integers}}$ and 3 columns that have $\color{green}{\text{datetime values}}$.

### Renaming the columns

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In the output of the code cell above, we are able to see that most of the column names are in camelcase and some of them are too long which makes it difficult to work with. Therefore, we will change the names of these columns from camelcase to snakecase

In [5]:
autos = autos.rename({'dateCrawled':'date_crawled',
                      'yearOfRegistration':'registration_year',
                      'monthOfRegistration':'registration_month',
                      'notRepairedDamage':'unrepaired_damage',
                      'dateCreated':'ad_created','offerType':'offer_type',
                      'abtest':'ab_test','vehicleType':'vehicle_type',
                      'gearBox':'gear_box','powerPS':'power_ps',
                      'fuelType':'fuel_type','nrOfPictures':'nr_of_pictures',
                      'postalCode':'postal_code',
                      'lastSeen':'last_seen'},axis = 1)

In [6]:
autos.head()


date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04             GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00             0.0      70435.0  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0.0      66954.0  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0.0      90480.0  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0.0      91074.0  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0.0      60437.0  2016-04-06 10:17:21

### Analyze the Columns

In [7]:
autos.describe(include='all')


date_crawled         name  seller offer_type         price  \
count                 42749        42749   42749      42749  4.274900e+04   
unique                41443        33631       1          2           NaN   
top     2016-03-21 12:38:30  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           86   42749      42748           NaN   
mean                    NaN          NaN     NaN        NaN  6.944650e+03   
std                     NaN          NaN     NaN        NaN  1.128076e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.199000e+03   
max                     NaN          NaN     NaN        NaN  1.234568e+07   

       ab_test vehicle_type  registration_year  gearbox      power_ps  model  \
count    42749        38300       42749.000000    40420  42749.000000  40381   
unique       2            8                NaN        2           NaN    244   
top       test    limousine                NaN  manuell           NaN   golf   
freq     22187        11008                NaN    31685           NaN   3390   
mean       NaN          NaN        2004.572271      NaN    113.181244    NaN   
std        NaN          NaN          95.278537      NaN    117.110608    NaN   
min        NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%        NaN          NaN        1999.000000      NaN     69.000000    NaN   
50%        NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%        NaN          NaN        2008.000000      NaN    150.000000    NaN   
max        NaN          NaN        9999.000000      NaN  10520.000000    NaN   

            kilometer  registration_month fuel_type       brand  \
count    42749.000000        42749.000000     38786       42748   
unique            NaN                 NaN         8          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     25545        9029   
mean    125614.400337            5.746427       NaN         NaN   
std      40044.593539            3.716406       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              34413                42748         42748.0  42748.000000   
unique                 2                   68             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               30160                 1674             NaN           NaN   
mean                 NaN                  NaN             0.0  50730.551324   
std                  NaN                  NaN             0.0  25858.624652   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30419.000000   
50%                  NaN                  NaN             0.0  49580.500000   
75%                  NaN                  NaN             0.0  71592.750000   
max                  NaN                  NaN             0.0  99996.000000   

                  last_seen  
count                 42748  
unique                34512  
top     2016-04-07 06:45:12  
freq                      7  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

- A few columns appear to have close to a single value like **'seller'**, **'offer_type'** and **'nr_of_pictures'**. We will remove these from our analysis.
- Some columns have numeric data which are stored as strings, for example: **'price'** and **'odometer'**. We will clean these and convert them into numeric data.
- The maximum value in the **'registration_year'** column is 9999 and the minimum value is 1000. The minimum values in the **'registration_month'** and **'power_ps'** columns is 0 which is impossible. We would need to investigate further for these columns. For now, we will clean the **'price'** and **'odometer'** columns.

### Clean the price and odometer columns

In [8]:
autos = autos.drop(columns = ['seller','offer_type','nr_of_pictures'])


In [9]:
autos = autos.rename({'kilometer':'odometer_km'},axis = 1)


In [10]:
autos['price'] = autos['price'].astype('str').str.replace('$','')
autos['price'] = autos['price'].astype('str').str.replace(',','').astype(int)
autos['odometer_km'] = autos['odometer_km'].astype('str').str.replace('km','')
autos['odometer_km'] = autos['odometer_km'].astype('str').str.replace(',','').astype(int)
autos

date_crawled                                              name  \
0      2016-03-24 11:52:17                                        Golf_3_1.6   
1      2016-03-24 10:58:45                              A5_Sportback_2.7_Tdi   
2      2016-03-14 12:52:21                    Jeep_Grand_Cherokee_"Overland"   
3      2016-03-17 16:54:04                               GOLF_4_1_4__3TÜRER   
4      2016-03-31 17:25:20                    Skoda_Fabia_1.4_TDI_PD_Classic   
...                    ...                                               ...   
42744  2016-03-16 16:06:20                             Alfa_Romeo_4C_1.8_TBi   
42745  2016-03-25 08:56:51                          BMW_X6_xDrive30d__1_HAND   
42746  2016-03-31 07:36:22                                          BMW_118i   
42747  2016-03-31 18:57:12  Mercedes_Benz_E_55_AMG_T_~19_Zoll~V_max~Tuev_Neu   
42748  2016-03-28 02:00:51     Mitsubishi_Lancer_Kombi_1.6_Comfort_AUTOMATIK   

       price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0        480     test          NaN               1993    manuell         0   
1      18300     test        coupe               2011    manuell       190   
2       9800     test          suv               2004  automatik       163   
3       1500     test   kleinwagen               2001    manuell        75   
4       3600     test   kleinwagen               2008    manuell        69   
...      ...      ...          ...                ...        ...       ...   
42744  56900     test        coupe               2016  automatik       241   
42745  31200  control    limousine               2009  automatik       235   
42746   9999     test    limousine               2008    manuell       143   
42747  15800  control        kombi               2003  automatik       500   
42748   1750     test        kombi               2004  automatik        98   

          model  odometer_km  registration_month fuel_type          brand  \
0          golf       150000                   0    benzin     volkswagen   
1           NaN       125000                   5    diesel           audi   
2         grand       125000                   8    diesel           jeep   
3          golf       150000                   6    benzin     volkswagen   
4         fabia        90000                   7    diesel          skoda   
...         ...          ...                 ...       ...            ...   
42744    andere         5000                   3    benzin     alfa_romeo   
42745   x_reihe        80000                   7    diesel            bmw   
42746       1er        70000                   9    benzin            bmw   
42747  e_klasse       150000                   9    benzin  mercedes_benz   
42748    lancer       150000                   8        be            NaN   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                   NaN  2016-03-24 00:00:00      70435.0  2016-04-07 03:16:57  
1                    ja  2016-03-24 00:00:00      66954.0  2016-04-07 01:46:50  
2                   NaN  2016-03-14 00:00:00      90480.0  2016-04-05 12:47:46  
3                  nein  2016-03-17 00:00:00      91074.0  2016-03-17 17:40:17  
4                  nein  2016-03-31 00:00:00      60437.0  2016-04-06 10:17:21  
...                 ...                  ...          ...                  ...  
42744              nein  2016-03-16 00:00:00      55129.0  2016-04-05 21:18:39  
42745              nein  2016-03-25 00:00:00      78532.0  2016-04-06 08:16:06  
42746              nein  2016-03-31 00:00:00      81671.0  2016-04-06 02:16:12  
42747              nein  2016-03-31 00:00:00      32105.0  2016-04-06 13:15:39  
42748               NaN                  NaN          NaN                  NaN  

[42749 rows x 17 columns]

In [11]:
print(autos['price'].unique().shape)
autos['price'].describe()

(2213,)


count    4.274900e+04
mean     6.944650e+03
std      1.128076e+05
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.199000e+03
max      1.234568e+07
Name: price, dtype: float64

In [12]:
autos['price'].value_counts().sort_index()

0           1273
1            129
2              3
3              1
5              4
            ... 
1250000        1
2995000        1
9999999        1
11111111       1
12345678       2
Name: price, Length: 2213, dtype: int64

- Some price values are abnormally low and high. Some are as low as 0 and some even reach 100 million. We will remove these outliers before we further investigate. We will set the minimum price to be 1000 and maximum 100000 dollars.

In [13]:
autos.loc[autos['price'].between(0,1000),'price'] = np.nan
autos.loc[autos['price'].between(100000,100000000),'price'] = np.nan
autos['price'].value_counts().sort_index()

1001.0      1
1025.0      1
1039.0      1
1049.0      7
1050.0     67
           ..
93000.0     1
95000.0     1
98500.0     1
99900.0     1
99999.0     3
Name: price, Length: 1920, dtype: int64

In [14]:
print(autos['price'].value_counts().shape)
autos['price'].describe()

(1920,)


count    32576.000000
mean      7180.439403
std       7934.030490
min       1001.000000
25%       2300.000000
50%       4490.000000
75%       8999.000000
max      99999.000000
Name: price, dtype: float64

Now that we have removed the outliers, we are left with around 38000 price entries. Standard deviation has also been reduced from around 480000 to almost 8000 which is a direct result of removing the exceedingly large and small price values.

In [15]:
autos['odometer_km'].value_counts(dropna=False)

150000    27683
125000     4395
100000     1837
90000      1381
80000      1302
70000      1178
60000       984
50000       878
5000        781
40000       749
30000       729
20000       626
10000       226
Name: odometer_km, dtype: int64

In [16]:
autos['odometer_km'].describe()

count     42749.000000
mean     125614.400337
std       40044.593539
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

The data in this column is fine and does not contain any outliers that could affect our analysis.

### Exploring the date columns

We are only interested in the number of sales or registrations on any single given day, therefore, we will not consider the time section for all three date columns.

In [17]:
date_crawled = autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index()*100


In [18]:
date_crawled


2016-03-05    2.502983
2016-03-06    1.450326
2016-03-07    3.581370
2016-03-08    3.324054
2016-03-09    3.405928
2016-03-10    3.251538
2016-03-11    3.188379
2016-03-12    3.618798
2016-03-13    1.511146
2016-03-14    3.520550
2016-03-15    3.244520
2016-03-16    2.902992
2016-03-17    3.164986
2016-03-18    1.347400
2016-03-19    3.616459
2016-03-20    3.574353
2016-03-21    3.590727
2016-03-22    3.204753
2016-03-23    3.291305
2016-03-24    2.987204
2016-03-25    3.279609
2016-03-26    3.225806
2016-03-27    3.059721
2016-03-28    3.569674
2016-03-29    3.464409
2016-03-30    3.317037
2016-03-31    3.298323
2016-04-01    3.483122
2016-04-02    3.469087
2016-04-03    3.876114
2016-04-04    3.892489
2016-04-05    1.342721
2016-04-06    0.290065
2016-04-07    0.152050
Name: date_crawled, dtype: float64

In [19]:
date_crawled.describe()


count    34.000000
mean      2.941176
std       0.972383
min       0.152050
25%       3.005333
50%       3.285457
75%       3.511193
max       3.892489
Name: date_crawled, dtype: float64

In [20]:
autos['date_crawled'].unique().shape


(41443,)

We see that there are 48213 unique values in the date_crawled column. All of which are from the months of April and March in the year 2016, with each value occupying almost 3% of the total values.

In [21]:
autos['ad_created'].str[:10].describe()


count          42748
unique            68
top       2016-04-03
freq            1674
Name: ad_created, dtype: object

In [22]:
ad_created = (autos['ad_created'].str[:10].value_counts(normalize=True).sort_index())*100


In [23]:
ad_created


2015-03-20    0.002339
2015-06-11    0.002339
2015-11-02    0.002339
2015-12-06    0.002339
2016-01-10    0.002339
                ...   
2016-04-03    3.915973
2016-04-04    3.897258
2016-04-05    1.230467
2016-04-06    0.287733
2016-04-07    0.149715
Name: ad_created, Length: 68, dtype: float64

In [24]:
ad_created.describe()


count    68.000000
mean      1.470588
std       1.615825
min       0.002339
25%       0.004679
50%       0.169599
75%       3.260386
max       3.915973
Name: ad_created, dtype: float64

In [25]:
autos['ad_created'].unique().shape


(69,)

The **ad_created** column contains $\color{green}{\text{76}}$ unique values with each unique value occupying about  $\color{green}{\text{1.3%}}$ on average of the total values. Which means that about $\color{green}{\text{650}}$ ads were created on a single day on average. Almost half of the values are from the months of March and April in the year 2016, while $\color{green}{\text{5-10}}$ values are from months in 2015.

In [26]:
autos['last_seen'].str[:10].value_counts()


2016-04-06    9296
2016-04-07    5726
2016-04-05    5390
2016-03-17    1200
2016-04-02    1110
2016-04-04    1099
2016-04-03    1042
2016-04-01    1031
2016-03-31    1028
2016-03-12    1007
2016-03-29    1006
2016-03-30     996
2016-03-28     959
2016-03-22     882
2016-03-21     862
2016-03-20     828
2016-03-25     822
2016-03-24     815
2016-03-23     805
2016-03-27     715
2016-03-26     714
2016-03-15     668
2016-03-16     665
2016-03-19     660
2016-03-11     588
2016-03-14     495
2016-03-10     480
2016-03-09     462
2016-03-08     342
2016-03-13     340
2016-03-18     284
2016-03-07     231
2016-03-06     155
2016-03-05      45
Name: last_seen, dtype: int64

In [27]:
autos['last_seen'].str[:10].describe()


count          42748
unique            34
top       2016-04-06
freq            9296
Name: last_seen, dtype: object

In [28]:
last_seen = (autos['last_seen'].str[:10].value_counts(normalize=True).sort_index())*100


In [29]:
last_seen.mean()


2.9411764705882355

In [30]:
autos['last_seen'].str[:10].unique().shape


(35,)

There are $\color{green}{\text{34}}$ unique date values in the **last_seen** column. All of which are from the year 2016 with each date occupying almost $\color{green}{\text{3%}}$, i.e., $\color{green}{\text{1500}}$ sales on average of all the values. Most sales are from the month of April, in which most of these are $\color{green}{\text{3}}$ consecutive days - The $\color{green}{\text{5th}}$, $\color{green}{\text{6th}}$ and $\color{green}{\text{7th}}$ of April. Which is almost $\color{green}{\text{4-7}}$ times the average sales number.

In [31]:
autos['registration_year'].describe()


count    42749.000000
mean      2004.572271
std         95.278537
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The **registration_year** column appears to have a number of invalid entries. The maximum value is 9999 and the minimum is 1000 which is impossible. We will need to fix these errors.

In [32]:
autos['registration_year'].unique()


array([1993, 2011, 2004, 2001, 2008, 1995, 1980, 2014, 1998, 2005, 1910,
       2016, 2007, 2009, 2002, 2018, 1997, 1990, 2017, 1981, 2003, 1994,
       1991, 1984, 2006, 1999, 2012, 2010, 2000, 1992, 2013, 1996, 1985,
       1989, 2015, 1968, 1982, 1976, 1983, 1959, 1973, 1111, 1969, 1971,
       1987, 1986, 1988, 1967, 1970, 1965, 1945, 1925, 1974, 1979, 1955,
       1978, 1972, 1977, 1961, 1963, 1964, 1960, 1966, 1975, 1937, 1936,
       5000, 1954, 1958, 9999, 1956, 3200, 1000, 1933, 1941, 1962, 8888,
       1500, 2200, 4100, 1929, 1951, 1957, 1940, 3000, 2066, 1949, 2019,
       1800, 1953, 1935, 1234], dtype=int64)

We will remove the years earlier than 1910 (this is the earliest plausible year after the invention of the first car). We shall also remove the years later than 2016 since a car cannot be first registered after it was last seen on the website (we saw earlier that all the values in the **last_seen** column are of the year 2016). We will not remove the entire row since we may lose valuable data, so we will assign the corresponding value as NaN.

In [33]:
autos.loc[autos['registration_year'] < 1910,'registration_year'] = np.nan


In [34]:
autos.loc[autos['registration_year'] > 2016, 'registration_year'] = np.nan


In [35]:
autos['registration_year'].describe()


count    41034.000000
mean      2002.803358
std          7.329715
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [36]:
(autos['registration_year'].value_counts(normalize=True).sort_index().tail(50))*100


1967.0    0.029244
1968.0    0.031681
1969.0    0.048740
1970.0    0.082858
1971.0    0.063362
1972.0    0.073110
1973.0    0.041429
1974.0    0.060925
1975.0    0.041429
1976.0    0.077984
1977.0    0.060925
1978.0    0.073110
1979.0    0.070673
1980.0    0.175464
1981.0    0.102354
1982.0    0.097480
1983.0    0.138909
1984.0    0.104791
1985.0    0.226641
1986.0    0.131598
1987.0    0.202271
1988.0    0.324121
1989.0    0.377736
1990.0    0.816396
1991.0    0.792026
1992.0    0.792026
1993.0    0.930935
1994.0    1.335478
1995.0    2.758688
1996.0    3.063313
1997.0    4.094166
1998.0    5.032412
1999.0    6.328898
2000.0    6.945460
2001.0    5.734269
2002.0    5.434518
2003.0    5.578301
2004.0    5.534435
2005.0    6.211922
2006.0    5.722084
2007.0    4.925184
2008.0    4.542574
2009.0    4.352488
2010.0    3.419116
2011.0    3.297266
2012.0    2.673393
2013.0    1.781450
2014.0    1.408588
2015.0    0.806648
2016.0    2.748940
Name: registration_year, dtype: float64

In [37]:
((autos['registration_year'].value_counts(normalize=True))*100).mean()


1.3157894736842102

Every year makes up approximately $\color{green}{\text{1.3%}}$ of the entire **registration_year** column, i.e., about $\color{green}{\text{620}}$ cars/year by average. About $\color{green}{\text{20-30}}$ cars/year were registered before the $\color{green}{\text{80s}}$ and the registrations start to increase after that. The maximum registrations were recorded in the year 2005 - around $\color{green}{\text{3000}}$, then we again see a steady decrease untill the year 2016 where the numbers jump from about $\color{green}{\text{400}}$ (in the year 2015) to almost $\color{green}{\text{1300}}$.

### Aggregating brand, price and odometer data¶


In [38]:
autos['brand'].describe()


count          42748
unique            40
top       volkswagen
freq            9029
Name: brand, dtype: object

In [39]:
(autos['brand'].value_counts())


volkswagen        9029
bmw               4692
opel              4649
mercedes_benz     4026
audi              3769
ford              2923
renault           2109
peugeot           1281
fiat              1093
seat               841
skoda              680
mazda              678
citroen            647
nissan             611
smart              591
toyota             529
sonstige_autos     472
hyundai            401
mini               388
volvo              384
mitsubishi         333
honda              319
kia                302
suzuki             279
porsche            264
alfa_romeo         249
chevrolet          200
chrysler           161
dacia              105
subaru              89
jeep                88
daihatsu            86
land_rover          85
trabant             73
jaguar              72
saab                60
rover               55
daewoo              55
lancia              51
lada                29
Name: brand, dtype: int64

There are $\color{green}{\text{40}}$ unique brands in the **brand column** with volkswagen being the most frequent with more than $\color{green}{\text{10000}}$ sales. We will leave aggregation of those brands whose frequencies are less than $\color{green}{\text{1%}}$ of the total values in the **brand column**.

In [40]:
brand_sel = (autos['brand'].value_counts(normalize=True))*100 > 1


In [41]:
brands = ((autos['brand'].value_counts(normalize=True))*100)[brand_sel]


In [42]:
brands


volkswagen        21.121456
bmw               10.975952
opel              10.875363
mercedes_benz      9.417984
audi               8.816787
ford               6.837747
renault            4.933564
peugeot            2.996631
fiat               2.556845
seat               1.967344
skoda              1.590718
mazda              1.586039
citroen            1.513521
nissan             1.429307
smart              1.382521
toyota             1.237485
sonstige_autos     1.104145
Name: brand, dtype: float64

In [43]:
sel_brands = brands.index


In [44]:
price_by_brand = {}

for b in sel_brands:
    mean_price = autos.loc[autos['brand'] == b, 'price'].mean()
    price_by_brand[b] = mean_price

In [45]:
price_by_brand

{'volkswagen': 6637.363463228831,
 'bmw': 9190.892596200682,
 'opel': 4285.673709733287,
 'mercedes_benz': 8935.777406789093,
 'audi': 9944.672738312083,
 'ford': 5270.623510292525,
 'renault': 3691.224960254372,
 'peugeot': 4093.9989235737353,
 'fiat': 3919.344680851064,
 'seat': 5935.990180032733,
 'skoda': 6994.877777777778,
 'mazda': 5539.8329809725155,
 'citroen': 4507.989648033126,
 'nissan': 5967.626506024097,
 'smart': 3774.7444444444445,
 'toyota': 5191.320166320166,
 'sonstige_autos': 13765.751612903226}

- Analysis of the results show that the top three most expenseive brands consists of **'Audi'** which has the highest average selling price of $\color{green}{\text{10242}}$ dollars, 2nd highest is **'Mercedes Benz'** at $\color{green}{\text{9202}}$ dollars and the 3rd highest average price is of **'BMW'** at $\color{green}{\text{8990}}$ dollars.
- The mid-tier priced brands are **'skoda'**, **'volkswagen'** and **'nissan'** with average prices of $\color{green}{\text{6894}}$, $\color{green}{\text{6652}}$ and $\color{green}{\text{6403}}$ dollars respectively.
- The **'sonstige_autos'** section which means **'other cars'** in English  displays the highest values but these may include newer or obsolete brands,therefore, we can ignore these in the results.
- And the cheapest brands are **'renault'**, **'smart'** and **'peugeot'** in increasing order of prices.

Next important parameter is the mileage of the cars. We have only the **odometer_km** column with us so we'd have to make do with it.

In [46]:
mileage = {}
for b in sel_brands:
    mean = autos.loc[autos['brand'] == b, 'odometer_km'].mean()
    mileage[b] = mean

In [47]:
mean_mileage_km = pd.Series(mileage)
mean_mileage_km = mean_mileage_km.sort_values(ascending = False)
mean_mileage_km

bmw               132966.751918
mercedes_benz     130715.350224
audi              129583.443884
opel              128613.680361
volkswagen        128606.157936
renault           128119.962067
mazda             125302.359882
ford              124324.324324
peugeot           123887.587822
citroen           121545.595054
seat              120939.357907
nissan            120932.896890
toyota            118809.073724
fiat              117758.462946
skoda             111610.294118
smart             100549.915398
sonstige_autos     84343.220339
dtype: float64

In [48]:
mean_price_dollars = pd.Series(price_by_brand)
mean_price_dollars = mean_price_dollars.sort_values(ascending = False)
mean_price_dollars 

sonstige_autos    13765.751613
audi               9944.672738
bmw                9190.892596
mercedes_benz      8935.777407
skoda              6994.877778
volkswagen         6637.363463
nissan             5967.626506
seat               5935.990180
mazda              5539.832981
ford               5270.623510
toyota             5191.320166
citroen            4507.989648
opel               4285.673710
peugeot            4093.998924
fiat               3919.344681
smart              3774.744444
renault            3691.224960
dtype: float64

In [49]:
df = pd.DataFrame(mean_price_dollars,columns = ['mean_price_dollars'])
df

mean_price_dollars
sonstige_autos        13765.751613
audi                   9944.672738
bmw                    9190.892596
mercedes_benz          8935.777407
skoda                  6994.877778
volkswagen             6637.363463
nissan                 5967.626506
seat                   5935.990180
mazda                  5539.832981
ford                   5270.623510
toyota                 5191.320166
citroen                4507.989648
opel                   4285.673710
peugeot                4093.998924
fiat                   3919.344681
smart                  3774.744444
renault                3691.224960

In [50]:
df['mean_mileage_km'] = mean_mileage_km
df = df.iloc[1:,:]
df['mean_price_dollars'] = df['mean_price_dollars'].astype(int)
df['mean_mileage_km'] = df['mean_mileage_km'].astype(int)

<ipython-input-50-82216bbf9c11>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['mean_price_dollars'] = df['mean_price_dollars'].astype(int)
<ipython-input-50-82216bbf9c11>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['mean_mileage_km'] = df['mean_mileage_km'].astype(int)


In [51]:
df

mean_price_dollars  mean_mileage_km
audi                         9944           129583
bmw                          9190           132966
mercedes_benz                8935           130715
skoda                        6994           111610
volkswagen                   6637           128606
nissan                       5967           120932
seat                         5935           120939
mazda                        5539           125302
ford                         5270           124324
toyota                       5191           118809
citroen                      4507           121545
opel                         4285           128613
peugeot                      4093           123887
fiat                         3919           117758
smart                        3774           100549
renault                      3691           128119

### Explore the  most common brand and/or model combinations¶

In [52]:
autos.loc[autos['brand']=='bmw','model'].describe().top

'3er'

In [53]:
d = {}
for b in autos['brand'].unique():
    d[b] = autos.loc[autos['brand']==b,'model'].describe().top
d
    

{'volkswagen': 'golf',
 'audi': 'a4',
 'jeep': 'grand',
 'skoda': 'octavia',
 'bmw': '3er',
 'peugeot': '2_reihe',
 'ford': 'fiesta',
 'mazda': '6_reihe',
 'nissan': 'micra',
 'renault': 'twingo',
 'mercedes_benz': 'c_klasse',
 'opel': 'corsa',
 'seat': 'ibiza',
 'citroen': 'andere',
 'honda': 'civic',
 'fiat': 'punto',
 'mini': 'cooper',
 'smart': 'fortwo',
 'hyundai': 'i_reihe',
 'sonstige_autos': nan,
 'alfa_romeo': '156',
 'subaru': 'impreza',
 'volvo': 'v40',
 'mitsubishi': 'colt',
 'kia': 'andere',
 'suzuki': 'andere',
 'lancia': 'ypsilon',
 'porsche': '911',
 'toyota': 'yaris',
 'chevrolet': 'andere',
 'dacia': 'logan',
 'daihatsu': 'cuore',
 'trabant': '601',
 'saab': 'andere',
 'chrysler': 'andere',
 'jaguar': 'andere',
 'daewoo': 'lanos',
 'rover': 'andere',
 'land_rover': 'freelander',
 'lada': 'niva',
 nan: nan}

In [54]:
df['most_common_model'] = pd.Series(d)

<ipython-input-54-68a91684a4fa>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['most_common_model'] = pd.Series(d)


In [55]:
df

mean_price_dollars  mean_mileage_km most_common_model
audi                         9944           129583                a4
bmw                          9190           132966               3er
mercedes_benz                8935           130715          c_klasse
skoda                        6994           111610           octavia
volkswagen                   6637           128606              golf
nissan                       5967           120932             micra
seat                         5935           120939             ibiza
mazda                        5539           125302           6_reihe
ford                         5270           124324            fiesta
toyota                       5191           118809             yaris
citroen                      4507           121545            andere
opel                         4285           128613             corsa
peugeot                      4093           123887           2_reihe
fiat                         3919           117758             punto
smart                        3774           100549            fortwo
renault                      3691           128119            twingo

### Explore the prices of  damaged and un-damaged cars¶

- We will rename the entries in the **unrepaired_damage** column since they are written in German

In [56]:
autos.loc[autos['unrepaired_damage']=='nein','unrepaired_damage'] = 'no'


In [57]:
autos.loc[autos['unrepaired_damage']=='ja','unrepaired_damage'] ='yes'


In [58]:
autos.loc[(autos['brand']=='bmw') & (autos['unrepaired_damage']=='no'), 'price'].describe()


count     3315.000000
mean      9906.334238
std       8539.016308
min       1050.000000
25%       3570.000000
50%       7550.000000
75%      13498.500000
max      81000.000000
Name: price, dtype: float64

In [61]:
d1 = {}
for b in autos['brand'].unique():
    d1[b] = (autos.loc[(autos['brand']==b) & (autos['unrepaired_damage']=='no'), 'price'].mean())

In [62]:
df['mean_price_without_damage'] = pd.Series(d1)

<ipython-input-62-adb9091db602>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['mean_price_without_damage'] = pd.Series(d1)


In [63]:
df

mean_price_dollars  mean_mileage_km most_common_model  \
audi                         9944           129583                a4   
bmw                          9190           132966               3er   
mercedes_benz                8935           130715          c_klasse   
skoda                        6994           111610           octavia   
volkswagen                   6637           128606              golf   
nissan                       5967           120932             micra   
seat                         5935           120939             ibiza   
mazda                        5539           125302           6_reihe   
ford                         5270           124324            fiesta   
toyota                       5191           118809             yaris   
citroen                      4507           121545            andere   
opel                         4285           128613             corsa   
peugeot                      4093           123887           2_reihe   
fiat                         3919           117758             punto   
smart                        3774           100549            fortwo   
renault                      3691           128119            twingo   

               mean_price_without_damage  
audi                        10911.243274  
bmw                          9906.334238  
mercedes_benz                9839.215625  
skoda                        7357.301095  
volkswagen                   7246.346596  
nissan                       6379.837758  
seat                         6176.149425  
mazda                        6128.812169  
ford                         5654.928475  
toyota                       5392.419192  
citroen                      4790.092838  
opel                         4566.008822  
peugeot                      4420.877717  
fiat                         4215.813102  
smart                        4044.146172  
renault                      3985.684318

In [83]:
d2 = {}
for b in autos['brand'].unique():
    d2[b] = (autos.loc[(autos['brand']==b) & (autos['unrepaired_damage']=='yes'), 'price'].mean())

In [84]:
df['mean_price_with_damage'] = pd.Series(d2)

<ipython-input-84-0cc0768a2817>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['mean_price_with_damage'] = pd.Series(d2)


In [85]:
autos.loc[(autos['unrepaired_damage'] != 'yes') & (autos['unrepaired_damage']!='no'),'unrepaired_damage'] = 'unknown'


In [86]:
autos['unrepaired_damage'].value_counts(dropna=False)


no         30160
unknown     8336
yes         4253
Name: unrepaired_damage, dtype: int64

In [87]:
df['mean_price_with_damage'] = df['mean_price_with_damage'].astype(int)


<ipython-input-87-26b4b8f8b3d4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['mean_price_with_damage'] = df['mean_price_with_damage'].astype(int)


In [88]:
df['non_damaged_vs_damaged_price_ratio'] = df['mean_price_without_damage']/df['mean_price_with_damage']


<ipython-input-88-280aa0ccb949>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['non_damaged_vs_damaged_price_ratio'] = df['mean_price_without_damage']/df['mean_price_with_damage']


In [89]:
df


mean_price_dollars  mean_mileage_km most_common_model  \
audi                         9944           129583                a4   
bmw                          9190           132966               3er   
mercedes_benz                8935           130715          c_klasse   
skoda                        6994           111610           octavia   
volkswagen                   6637           128606              golf   
nissan                       5967           120932             micra   
seat                         5935           120939             ibiza   
mazda                        5539           125302           6_reihe   
ford                         5270           124324            fiesta   
toyota                       5191           118809             yaris   
citroen                      4507           121545            andere   
opel                         4285           128613             corsa   
peugeot                      4093           123887           2_reihe   
fiat                         3919           117758             punto   
smart                        3774           100549            fortwo   
renault                      3691           128119            twingo   

               mean_price_without_damage  mean_price_with_damage  \
audi                        10911.243274                    5289   
bmw                          9906.334238                    5676   
mercedes_benz                9839.215625                    4426   
skoda                        7357.301095                    4955   
volkswagen                   7246.346596                    4161   
nissan                       6379.837758                    3130   
seat                         6176.149425                    3296   
mazda                        6128.812169                    2823   
ford                         5654.928475                    3218   
toyota                       5392.419192                    3804   
citroen                      4790.092838                    2871   
opel                         4566.008822                    3037   
peugeot                      4420.877717                    2850   
fiat                         4215.813102                    3440   
smart                        4044.146172                    1967   
renault                      3985.684318                    2245   

               non_damaged_vs_damaged_price_ratio  
audi                                     2.063007  
bmw                                      1.745302  
mercedes_benz                            2.223049  
skoda                                    1.484824  
volkswagen                               1.741492  
nissan                                   2.038287  
seat                                     1.873832  
mazda                                    2.171028  
ford                                     1.757280  
toyota                                   1.417566  
citroen                                  1.668441  
opel                                     1.503460  
peugeot                                  1.551185  
fiat                                     1.225527  
smart                                    2.055997  
renault                                  1.775360

In [90]:
df['non_damaged_vs_damaged_price_ratio'].mean()


1.7684772772337447

### Conclusion

- The top three most expensive brands are **'Audi'**, **'Mercedes Benz'** and **'BMW'**. 


- **'Skoda'**, **'Ford'** , **'Nissan'**, **'Volkswagen'** and **'Toyota'** are relatively less expensive and among the cheapest ones are **'Renault'**, **'Smart'** and **'Fiat'**.

There seems to be no direct correlation between mileage and average prices.

The average price of cars without damage is about $\color{green}{\text{1.78}}$ times the average price of non-damaged cars which is not surprising. For most cases this ratio is greater than $\color{green}{\text{2}}$.